node/src/components/linear_chain_sync/operations.rs

```diff
@@ -33,24 +30,23 @@ const TIMEOUT_DURATION: Duration = Duration::from_millis(100);
 /// Fetches an item. Keeps retrying to fetch until it is successful. Assumes no integrity check is
 /// necessary for the item. Not suited to fetching a block header or block by height, which require
 /// verification with finality signatures.
-async fn fetch_retry_forever<T, REv, I>(
-    effect_builder: EffectBuilder<REv>,
+async fn fetch_retry_forever<T>(
+    effect_builder: EffectBuilder<JoinerEvent>,
     id: T::Id,
-) -> Result<Box<T>, FetcherError<T, I>>
+) -> Result<Box<T>, FetcherError<T, NodeId>>
 where
     T: Item + 'static,
-    REv: From<FetcherRequest<I, T>> + From<NetworkInfoRequest<I>>,
-    I: Eq + Debug + Clone + Send + 'static,
+    JoinerEvent: From<FetcherRequest<NodeId, T>>,
 {
     loop {
-        for peer in effect_builder.get_peers_in_random_order().await {
+        'peer_loop: for peer in effect_builder.get_peers_in_random_order().await {
```

Comment:
This will request the same elements from every peer we are connected to. Was that intentional? If so, I don't think it's a good decision.